beacon_node/beacon_chain/src/block_verification.rs

```diff
@@ -729,19 +784,6 @@ impl<T: BeaconChainTypes> GossipVerifiedBlock<T> {
             return Err(BlockError::BlockIsAlreadyKnown);
         }
 
-        // Check that we have not already received a block with a valid signature for this slot.
-        if chain
-            .observed_block_producers
-            .read()
-            .proposer_has_been_observed(block.message())
-            .map_err(|e| BlockError::BeaconChainError(e.into()))?
-        {
-            return Err(BlockError::RepeatProposal {
-                proposer: block.message().proposer_index(),
-                slot: block.slot(),
-            });
-        }
```

Comment:
Context for the 2nd reviewer, we have to remove this check for blocks which equivocate, or we risk "forgetting" about some equivocations. If this check were in place for equivocating blocks, a flow which defeats it is:

1. Block B is gossip verified, cache contains `block_roots = {B}` for the current slot & proposer index.
2. Equivocating block `A` attempts gossip verification, and is rejected by this "read-only" check, because we can see that `A` is an equivocation with respect to `{B}`. However we then _forget_ about `A` by returning an error _without updating the set of seen block roots_.
3. Block B does a "late" equivocation check against the cache before it is broadcast. The cache still only contains `{B}` (no `A`) so it doesn't think there's been an equivocation so block B gets incorrectly broadcast.

We could change the semantics of `proposer_has_been_observed` so that it _doesn't_ return an error when `block_roots = {B}` and we are seeing `A`, but I think this is quite counterintuitive and could prove dangerous in other places where the function is used (e.g. when detecting whether an RPC block is an equivocation).

If we wanted to keep some ability to reject duplicates here we could return an error _only for duplicates_, like:

```rust
        if chain
            .observed_block_producers
            .read()
            .proposer_has_been_observed(block.message())
            .map_err(|e| BlockError::BeaconChainError(e.into()))?
            .is_duplicate()
        {
            return Err(BlockError::RepeatProposal {
                proposer: block.message().proposer_index(),
                slot: block.slot(),
            });
        }
```

However, most duplicates should already have been caught by the fork choice check, and if they slip through, they'll be caught when we call `observe_proposal` after signature verification. So we only risk running one extra signature verification.